tutorials/02_functions-types.md

```diff
@@ -0,0 +1,518 @@
+
+```haskell
+data MyType a b = MyTypeC1 a Int | MyTypeC2 String b | MyType3
+                deriving Show
+
+x :: MyType Bool Float
+x = MyTypeC1 True 10
+```
+
+Usually, when there is just one data constructor, the name is the same as of the type constructor (but it is not a rule).
+
+### Record types
+
+Imagine you want to create a type for `Person` which contains `name`, `age`, `gender`, and `city`. You would need to do:
+
+```haskell
+data Gender = Male | Female
+            deriving Show
+
+data Person = Person String Int Gender String
+            deriving Show
+
+name :: Person -> String
+name (Person x _ _ _) = x
+
+age :: Person -> Int
+age (Person _ x _ _) = x
+
+gender :: Person -> Gender
+gender (Person _ _ x _) = x
+
+city :: Person -> String
+city (Person _ _ _ x) = x
+```
+
+Not very nice, right?! And we have just 4 attributes of a person. Luckily there is syntactic sugar which makes it easier for us called record:
+
+```haskell
+data Gender = Male | Female
+            deriving Show
+
+data Person = Person
+            { name   :: String
+            , age    :: Int
+            , gender :: Gender
+            , city   :: String
+            } deriving Show
+```
+
+Now try to create a type `Pet` which also contains `name` and `age`. You will get an error which is logical, you cannot have two functions with the same name! One option is to rename it to `namePerson` and `namePet`, the second is available to you only if you have GHC 8.0.1 or higher and it is with language extension [DuplicateRecordFields](https://downloads.haskell.org/~ghc/master/users-guide/glasgow_exts.html#duplicate-record-fields):
+
+```haskell
+{-# LANGUAGE DuplicateRecordFields #-}
+
+{-
+  Seven Deadly Sins ordered by Dante Alighieri
+  see: https://simple.wikipedia.org/wiki/Seven_deadly_sins
+-}
+data Sin = Lust | Gluttony | Greed | Sloth | Wrath | Envy | Pride
+         deriving (Show, Read, Eq, Ord, Enum, Bounded)
+
+data Gender = Male | Female
+            deriving (Show, Read, Eq)      -- Why not Ord? Gender equality!
+
+data Person = Person
+            { name   :: String
+            , age    :: Int
+            , gender :: Gender
+            , city   :: String
+            } deriving (Show, Read, Eq)
+
+data Pet = Pet
+         { name   :: String
+         , age    :: Int
+         } deriving (Show, Read)
+
+olderPet :: Pet -> Pet
+olderPet pet = pet { age = (age pet + 1) }
+```
+
+You can also see that there is a shorthand for updating the value of record - creating new edited record from previous. Now you can try some derived behaviour from typeclasses such as `show`, `read`, or `==`:
```

Comment:
This is one of the weaker parts of Haskell: it actually does NOT have records (as explained, there are just syntactic sugar).  Newer languages from the Haskell family PureScript and Elm elaborated on this and introduced 'proper' records.